lib_test/idl_test_common.ml

```diff
@@ -0,0 +1,254 @@
+(*
+ * Copyright (C) 2018 Citrix Inc
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+let write_str filename str =
+  let oc = open_out filename in
+  Printf.fprintf oc "%s" str;
+  close_out oc
+
+let read_str filename =
+  let ic = open_in filename in
+  let n = in_channel_length ic in
+  let s = Bytes.create n in
+  really_input ic s 0 n;
+  close_in ic;
+  s
+
+open Idl
+
+module type PATHS = sig
+  val test_data_path : string
+  (** Path under which we look for or generate requests and responses. For example,
+      if test_data_path = 'foo', this module will search for or generate requests
+      matching 'foo/requests/<RPC name>.request.<n>' and responses matching
+      'foo/responses/<RPC name>.response.<n>' *)
+end
+
+(** The following module implements test cases that write test
+    RPC requests and responses in JSON that can be used to
+    verify that subsequent versions of an API can still parse
+    them.
+
+    The test cases are obtained by obtaining the implementation
+    of the module generated when applying the API functor to
+    this module.
+
+    The test data will be written to the path specified in the
+    PATH module passed in *)
+module GenTestData (P:PATHS) = struct
+  type implementation = unit Alcotest.test_case list ref
+
+  let tests : unit Alcotest.test_case list ref = ref []
+  let description = ref None
+
+  let implement x = description := Some x; tests
+
+  type ('a,'b) comp = 'a
+  type 'a res = unit
+  type _ fn =
+    | Function : 'a Idl.Param.t * 'b fn -> ('a -> 'b) fn
+    | Returning : ('a Idl.Param.t * 'b Idl.Error.t) -> ('a, _) comp fn
+
+  let returning a err = Returning (a, err)
+  let (@->) = fun t f -> Function (t, f)
+
```

Comment:
Some of this code is from ocaml-rpc, why duplicate it here instead of `let returning =  Idl.returning` etc.?

Reply:
The types are different unfortunately.